Cue context menu for scrolling waveform

hacksdump opened the pull request:
Adds context menu to hot cues on the scrolling waveform just like WOverview. Right-click on the label to invoke the context menu.

TODOs:
- [x]  Disallow opening context menu when a track is playing.

- [x]  Open context menu when right-clicking in the vicinity of marker line.

- [x]  Highlight cue mark on hover.

hacksdump committed:
wwaveformviewer: add context menu to scrolling waveform

hacksdump committed:
wwaveformviewer: limit to cue markers on screen

hacksdump committed:
wwaveformviewer: handle menu close

hacksdump committed:
wwaveformviewer: add menu to vertical orientation

hacksdump committed:
wwaveformrendermark: remove unnecessary state variable

uklotzde commented:
Please rebase and retarget on master. The commit history contains unrelated commits.

hacksdump commented:
Should I rebase to mixxxdj:2.3 or master?

uklotzde commented:
@hacksdump Not sure. I just saw the unrelated commits and then later noticed that @Be-ing retargeted the PR to 2.3 which might have caused them.

Since this is a new "feature" master would be more appropriate imo.

daschuer commented:
Yes, new feature=master. I have changed the target branch.

Be-ing commented:
This was already planned to go to 2.3 before cutting the 2.3 branch.

daschuer commented:
We need to get used to the new dead line based release strategy. 
"Only one last feature" was the source of the delay of 2.3. 

2.2 was release 11 Jan 2019. If we would have stopped to add features to 2.3, it would have been release in Juli 2019. Now we would have had released 2.4, a 2.5 beta and we where considering shiny new features for 2.6.

Be-ing commented:
I think we should have a discussion about why 2.3 took so long. But this PR is not the place for that.

uklotzde commented:
Try to rebase *onto* 2.3. If this works without conflicts we could decide after the review if it is safe to merge into 2.3.

hacksdump committed:
wwaveformviewer: enable mouse tracking

hacksdump committed:
wwaveformviewer: stop unnecessary bending after menu close

uklotzde commented on src/waveform/renderers/waveformwidgetrenderer.cpp:
```diff
@@ -288,3 +288,31 @@ void WaveformWidgetRenderer::setTrack(TrackPointer track) {
+    CuePointer pSelectedCue;
+    QList<CuePointer> cueList = getTrackInfo()->getCuePoints();
+    for (const auto& pCue : cueList) {
+        if (pCue->getHotCue() == pSelectedMark->getHotCue()) {
```
Do we need to skip *cue points* with type != mixxx::CueType::HotCue? pCue->getHotCue() is -1 for many other types.

Maybe like so:
```
VERIFY_OR_DEBUG_ASSERT(pCue->getType() == mixxx::CueType::HotCue) {
    continue;
}
```

Holzhaus replied:
Due to the Serato/Rekordbox import loop cues can also have a hotcue number now.

uklotzde replied:
Then we need to compare both type and number here? Are there any rules regarding uniqueness of this number in relation to the type?

I haven't checked yet how the remaining code deals with this situation.

Holzhaus replied:
The numbers need to be unique or strange things might happen. Currently the only way to get loop cues with numbers is via import from Rekordbox/Serato, so there shouldn't be an issue.

hacksdump replied:
> Do we need to skip _cue points_ with type != mixxx::CueType::HotCue? pCue->getHotCue() is -1 for many other types.

Currently, only hotcues seem to be having a context menu.

uklotzde commented on src/waveform/renderers/waveformwidgetrenderer.cpp:
```diff
@@ -288,3 +288,31 @@ void WaveformWidgetRenderer::setTrack(TrackPointer track) {
     }
 }
+
+std::optional<CuePointer> WaveformWidgetRenderer::getCueAtPoint(QPoint point) {
```
Is wrapping a pointer into std::optional common practice in C++? I guess in this case returning a nullptr is appropriate and simplifies the code. Checking for both nullopt and nullptr feels overcomplicated.

uklotzde commented on src/waveform/renderers/waveformwidgetrenderer.h:
```diff
@@ -93,6 +96,10 @@ class WaveformWidgetRenderer {
     }
 
     void setTrack(TrackPointer track);
+    inline void setMarkLabelBoundaries(
```
inline is redundant

uklotzde commented on src/waveform/waveformmarklabel.h:
```diff
@@ -27,6 +27,10 @@ class WaveformMarkLabel {
         return m_areaRect;
     };
 
+    inline void setAreaRect(const QRectF& areaRect) {
```
inline is redundant

uklotzde commented on src/widget/wcuemenupopup.h:
```diff
@@ -59,6 +54,8 @@ class WCueMenuPopup : public QWidget {
     void slotChangeCueColor(mixxx::RgbColor::optional_t color);
 
   private:
+    void closeEvent(QCloseEvent* event) override;
```
This method is protected in the base class and you should use the same visibility here. Reducing the visibility in derived classes is not recommended, because it does not work as intended.

uklotzde commented on src/widget/wwaveformviewer.cpp:
```diff
@@ -77,6 +83,13 @@ void WWaveformViewer::mousePressEvent(QMouseEvent* event) {
+        if (cueAtClickPos) {
+            m_pCueMenuPopup->setTrackAndCue(currentTrack, *cueAtClickPos);
+            m_pCueMenuPopup->popup(event->globalPos());
+            m_bHotcueMenuShowing = true;
```
Why do we need a separate boolean for tracking the status instead of using QWidget::isVisible()?

hacksdump committed:
wwaveformviewer: detect right click in mark line vicinity

hacksdump committed:
waveformwidgetrenderer: remove std::optional wrapping

hacksdump committed:
wcuemenupopup: override closeEvent as protected

hacksdump committed:
remove redundant inlines

hacksdump committed:
remove redundant menu open state variable

hacksdump committed:
wwaveformviewer: highlight hovered mark

hacksdump committed:
wwaveformviewer: disable cue menu when playing

hacksdump committed:
use functor syntax for connect calls

hacksdump committed:
waveformrendermark: cleaner rectangle initialisation

Be-ing commented on src/waveform/renderers/waveformrendermark.cpp:
```diff
@@ -31,7 +31,8 @@ void WaveformRenderMark::setup(const QDomNode& node, const SkinContext& context)
 void WaveformRenderMark::draw(QPainter* painter, QPaintEvent* /*event*/) {
     PainterScope PainterScope(painter);
-
+    // Maps mark image offset to mark object.
+    QMap<WaveformMarkPointer, int> marksOnScreen;
```
This comment wouldn't be necessary with a better name for the variable. How about `marksToScreenPositions`?

hacksdump replied:
Renaming this variable should be better. But the positions are relative to the scrolling widget, not the screen. So it may be renamed to `marksToWidgetPositions`. But the word `screen` indicates that this `QMap` only stores marks visible on screen.

In the overview widget, the size of the mark image is equal to the widget, thus, this position information is not required to detect if a mouse pointer is near/on the marker.
But I wouldn't suggest using the complete widget area just to draw markers on the scrolling waveform since that would take >10x canvas space required. So we will keep sending position with marks as a `QMap`.

Be-ing replied:
Okay, then change the comment to
```suggestion
    // Maps mark object to its position in the widget
    QMap<WaveformMarkPointer, int> marksOnScreen;
```

Be-ing commented on src/waveform/renderers/waveformrendermark.cpp:
```diff
@@ -69,23 +69,29 @@ void WaveformRenderMark::draw(QPainter* painter, QPaintEvent* /*event*/) {
 
                 // Check if the current point need to be displayed
```
```suggestion
                // Check if the current point needs to be displayed
```
Please fix this typo in the existing comment

Be-ing commented on src/waveform/renderers/waveformwidgetrenderer.h:
```diff
@@ -39,6 +40,8 @@ class WaveformWidgetRenderer {
 
     inline const char* getGroup() const { return m_group;}
     const TrackPointer getTrackInfo() const { return m_pTrack;}
+    // Get cue at a point on the waveform widget.
```
```suggestion
    /// Get cue at a point on the waveform widget.
```
We're using `///` for documentation so this is picked up by IDE tooltips and Doxygen

Be-ing commented on src/waveform/renderers/waveformwidgetrenderer.h:
```diff
@@ -93,6 +96,9 @@ class WaveformWidgetRenderer {
     }
 
     void setTrack(TrackPointer track);
+    void setMarkLabelBoundaries(QMap<WaveformMarkPointer, int> markLabelOffsets) {
+        m_markLabelOffsets = markLabelOffsets;
+    }
```
"Offset" implies that these are somehow shifted from another position.
```suggestion
    void setMarkLabelPositions(QMap<WaveformMarkPointer, int> markLabelPositions) {
        m_markLabelPositions = markLabelPositions;
    }
```

Be-ing commented on src/waveform/renderers/waveformwidgetrenderer.cpp:
```diff
@@ -288,3 +288,48 @@ void WaveformWidgetRenderer::setTrack(TrackPointer track) {
+            return pMark;
+        }
+    }
+    return static_cast<WaveformMarkPointer>(nullptr);
```
```suggestion
    return nullptr;
```
The explicit casting isn't necessary and I don't think it makes the code any easier to understand.

Be-ing commented on src/widget/wwaveformviewer.cpp:
```diff
@@ -208,13 +246,38 @@ void WWaveformViewer::setPlayMarkerPosition(double position) {
+            }
+        }
+    }
+    return static_cast<CuePointer>(nullptr);
```
```suggestion
    return nullptr;
```

Be-ing commented on src/widget/wwaveformviewer.cpp:
```diff
@@ -77,18 +79,29 @@ void WWaveformViewer::mousePressEvent(QMouseEvent* event) {
+            auto cueAtClickPos = getCuePointerFromCueMark(m_pHoveredMark);
+            if (cueAtClickPos) {
+                m_pCueMenuPopup->setTrackAndCue(currentTrack, cueAtClickPos);
+                m_pCueMenuPopup->popup(event->globalPos());
```
This is affected by the same bug as #2663, but don't worry about that until we merge #2663.

hacksdump replied:
>  We can use the same general approach to add right click menus for beats, downbeats, phrases, and section markers. Hopefully we can reuse some of this code too.

This functionality to find if a point lies in/near a cue marker, beat, downbeat, phrase marker will be quite common. Can this be abstracted to a new parent class?
For this PR, I was thinking of moving this functionality to `WaveformMark` so that it can be commonly used by `WOverview` and `WaveformWidgetRenderer`.

Be-ing commented:
Good work! The color change when hovering the mark is a nice detail that makes this feature easier to discover. There are a few minor ways to make the code a little easier to understand, but otherwise the code looks good. We can use the same general approach to add right click menus for beats, downbeats, phrases, and section markers. Hopefully we can reuse some of this code too.

uklotzde commented on src/widget/wwaveformviewer.cpp:
```diff
@@ -208,13 +246,38 @@ void WWaveformViewer::setPlayMarkerPosition(double position) {
+
+CuePointer WWaveformViewer::getCuePointerFromCueMark(WaveformMarkPointer pMark) {
+    if (pMark) {
+        QList<CuePointer> cueList = m_waveformWidget->getTrackInfo()->getCuePoints();
```
I suggest to add the following debug assertion:

```
DEBUG_ASSERT(pMark->getHotCue() != Cue::kNoHotCue);
```

Just to ensure that the mark always has a number that uniquely identifies a `Cue` object.

Be-ing replied:
This will fail currently. I can right click on loop markers with no hotcue number, which should not happen.

uklotzde commented:
Code is almost ready, only some minor comments. This feature would be a nice addition for 2.3 to complete the cue improvements and extensions. Let's keep 2.3 as the target.

uklotzde commented:
wwaveformviewer.h seems to be missing an #include directive? The SCons builds fail.

hacksdump commented:
>  We can use the same general approach to add right click menus for beats, downbeats, phrases, and section markers. Hopefully we can reuse some of this code too.

This functionality to find if a point lies in/near a cue marker, beat, downbeat, phrase marker will be quite common. Can this be abstracted to a new parent class?
For this PR, I was thinking of moving this functionality to `WaveformMark` so that it can be commonly used by `WOverview` and `WaveformWidgetRenderer`. Suggestions?

Be-ing commented:
>  For this PR, I was thinking of moving this functionality to WaveformMark so that it can be commonly used by WOverview and WaveformWidgetRenderer. Suggestions?

Seems like a good idea to me.

hacksdump committed:
waveformmark: add function to detect point in mark

hacksdump committed:
rename variables and change comments

hacksdump committed:
wwaveformviewer: limit cue menu to hotcues

Be-ing commented on src/waveform/renderers/waveformwidgetrenderer.cpp:
```diff
@@ -290,44 +290,15 @@ void WaveformWidgetRenderer::setTrack(TrackPointer track) {
-            markLineVicinity = QRect(markLineOffset - lineHoverpadding,
-                    0,
-                    lineHoverpadding * 2,
-                    getHeight());
```
So much less code. Beautiful :)

Be-ing commented on src/widget/woverview.cpp:
```diff
@@ -453,16 +451,7 @@ void WOverview::mouseMoveEvent(QMouseEvent* e) {
-                pMark->m_linePosition >= hoveredPosition - lineHoverPadding && pMark->m_linePosition <= hoveredPosition + lineHoverPadding;
-
-        if (pMark->m_label.area().contains(e->pos()) || lineHovered) {
+        if (pMark->contains(e->pos(), m_orientation)) {
```
Thanks for consolidating the code here in WOverview too.

Be-ing commented:
Fix the scons build and rebase onto 2.3, then this will be ready to merge.

uklotzde commented on src/waveform/renderers/waveformmark.cpp:
```diff
@@ -99,7 +99,24 @@ WaveformMark::WaveformMark(const QString& group,
+    bool lineHovered = m_linePosition >= position - lineHoverPadding &&
+            m_linePosition <= position + lineHoverPadding;
+
+    return (m_label.area().contains(point) || lineHovered);
```
Unneeded round brackets

uklotzde commented on src/waveform/renderers/waveformmark.h:
```diff
@@ -66,6 +66,9 @@ class WaveformMark {
     }
 
+    // Check if a point (in image co-ordinates) lies on drawn image.
+    bool contains(QPoint point, Qt::Orientation orientation);
```
This method should be const

uklotzde commented on src/waveform/renderers/waveformwidgetrenderer.cpp:
```diff
@@ -288,3 +288,19 @@ void WaveformWidgetRenderer::setTrack(TrackPointer track) {
+        QPoint pointInImageSpace;
+        if (getOrientation() == Qt::Horizontal) {
+            pointInImageSpace = QPoint(point.x() - markImagePositionInWidgetSpace, point.y());
+        } else { /* Vertical */
```
You could replace this comment with `DEBUG_ASSERT(getOrientation() == Qt::Vertical);` to document **and** verify your assumptions.

uklotzde commented on src/waveform/renderers/waveformwidgetrenderer.h:
```diff
@@ -39,6 +40,8 @@ class WaveformWidgetRenderer {
     inline const char* getGroup() const { return m_group;}
     const TrackPointer getTrackInfo() const { return m_pTrack;}
+    /// Get cue mark at a point on the waveform widget.
+    WaveformMarkPointer getCueMarkAtPoint(QPoint point);
```
The method should be const

uklotzde commented on src/waveform/renderers/waveformwidgetrenderer.h:
```diff
@@ -151,6 +157,7 @@ class WaveformWidgetRenderer {
 private:
     DISALLOW_COPY_AND_ASSIGN(WaveformWidgetRenderer);
     friend class WaveformWidgetFactory;
+    QMap<WaveformMarkPointer, int> m_markPositions;
```
In a later version we might think about a better encapsulation of this QMap. We use this type at two different places with different names ("m_marksOnScreen" vs. "m_markPositions"). You don't need to change this now.

uklotzde commented on src/widget/wwaveformviewer.h:
```diff
@@ -69,13 +72,21 @@ private slots:
 
     friend class WaveformWidgetFactory;
+
+    CuePointer getCuePointerFromCueMark(WaveformMarkPointer pMark);
```
Method should be const

uklotzde commented on src/widget/wwaveformviewer.h:
```diff
@@ -69,13 +72,21 @@ private slots:
+    CuePointer getCuePointerFromCueMark(WaveformMarkPointer pMark);
+    void highlightMark(WaveformMarkPointer pMark);
+    void unhighlightMark(WaveformMarkPointer pMark);
+    bool isPlaying();
```
Method should be const

hacksdump committed:
wwaveformviewer: fix scons build

uklotzde commented:
The SCons builds are still failing: https://travis-ci.org/github/mixxxdj/mixxx/builds/687546806

hacksdump committed:
convert to const methods

hacksdump commented:
Rebased onto 2.3

Be-ing commented:
Somehow there is a conflict in src/util/logging.cpp since the rebase.

hacksdump commented:
The base branch for the PR needs to be changed now.

Be-ing commented:
You can do that yourself. Click the Edit button at the top of the page.

hacksdump commented:
Still contains commits by other authors.

Be-ing commented:
Do `git rebase -i 2.3` and delete the commits that aren't yours.
https://git-scm.com/book/en/v2/Git-Branching-Rebasing

uklotzde commented:
Try: `git rebase --onto 2.3 <hash-of-your-1st-commit>^`

Please note the *circumflex* character after the commit hash.

uklotzde commented:
i.e. `git rebase --onto 2.3 8b9cfed18cec1ad57c880a056a75bc1bde3969c4^`

Holzhaus commented:
Or `<hash>~1` which also works on windows.

hacksdump commented:
Thanks. I just deleted other commits. Would have been easier to just rebase to my first commit. Fixed the scons build as well. Let's wait for CI builds.

uklotzde commented:
> Or `<hash>~1` which also works on windows.

Of course, both ^ and ~1 are equivalent.

uklotzde commented:
Works, nice feature :+1:

Be-ing commented:
There is a merge conflict since #2663 was merged. Fix that and also please use the new utility function introduced in that PR for showing CueMenuPopup in the scrolling waveform.

uklotzde commented:
I have merged #2663. A minor merge conflict needs to be resolved and we could use mixxx::widgethelper::mapPopupToScreen() for this new popup menu.

uklotzde commented:
Race condition while commenting on this PR ;)

hacksdump commented:
> we could use mixxx::widgethelper::mapPopupToScreen() for this new popup menu.

Writing this explicitly for every child of QMenu seems cumbersome. Every menu needs to be within mixxx boundaries.
I was thinking something along the lines of creating a new QMenu derivative which redefines the popup function and then using this class instead of QMenu to create our menus.

For this PR, I will just change `WWaveformViewer` in similar manner to `WOverview`.

hacksdump committed:
resolve merge conflict

Be-ing commented on src/widget/wwaveformviewer.cpp:
```diff
@@ -83,7 +85,11 @@ void WWaveformViewer::mousePressEvent(QMouseEvent* event) {
+                        windowHandle()->screen()->size(),
+                        event->globalPos(),
+                        m_pCueMenuPopup->size());
+                m_pCueMenuPopup->popup(cueMenuTopLeft);
```
You don't need to bother going back and changing this commit, but for the future please make merge commits the bare minimum required to resolve conflicts and get it to build and run, then add new changes in following commits.

hacksdump replied:
Got that. Shouldn't have combined it into the merge resolution commit. Will keep that in mind.

uklotzde commented:
Pre-commit warnings are unrelated. LGTM.

Do we agree to merge this into 2.3?

uklotzde commented:
Great. Thank you for contributing these UX improvements, Harshit! Working on this PR was a pleasure and also a constructive collaboration among everyone involved.
